js/src/logger.ts

```diff
@@ -1,18 +1,172 @@
 import axios, { AxiosInstance } from "axios";
 import { v4 as uuidv4 } from "uuid";
 
-import iso from "./isomorph";
+import iso, { IsoAsyncLocalStorage, CallerLocation } from "./isomorph";
+import { runFinally } from "./util";
 
-interface BraintrustState {
-  current_project: Project | null;
-  current_experiment: Experiment | null;
-}
+export type SetCurrentArg = { setCurrent?: boolean };
 
-let _state: BraintrustState = {
-  current_project: null,
-  current_experiment: null,
+export type StartSpanArgs = {
+  name: string;
+  spanAttributes?: Record<any, any>;
+  startTime?: number;
+  event?: ExperimentLogPartialArgs & Partial<IdField>;
 };
 
+export type StartSpanOptionalNameArgs = Omit<StartSpanArgs, "name"> &
+  Partial<Pick<StartSpanArgs, "name">>;
+
+/**
+ * A Span encapsulates logged data and metrics for a unit of work. This interface is shared by all span implementations.
+ *
+ * We suggest using one of the various `startSpan` methods, instead of creating Spans directly. See `Span.startSpan` for full details.
+ */
+export interface Span {
+  // Row ID of the span.
+  id: string;
+
+  // Span ID of the span. This is used to link spans together.
+  span_id: string;
+
+  // Span ID of the root span in the full trace.
+  root_span_id: string;
+
+  /**
+   * Incrementally update the current span with new data. The event will be batched and uploaded behind the scenes.
+   *
+   * @param event: Data to be logged. See `Experiment.log` for full details.
+   */
+  log(event: ExperimentLogPartialArgs): void;
+
+  /**
+   * Create a new span. This is useful if you want to log more detailed trace information beyond the scope of a single log event. Data logged over several calls to `Span.log` will be merged into one logical row.
+   *
+   * We recommend running spans within a callback (using `startSpanWithinCallback`) to automatically mark them as current and ensure they are terminated. If you wish to start a span outside a callback, be sure to terminate it with `span.end()`.
+   *
+   * @param args.name The name of the span.
+   * @param args.span_attributes Optional additional attributes to attach to the span, such as a type name.
+   * @param args.start_time Optional start time of the span, as a timestamp in seconds.
+   * @param args.event Data to be logged. See `Experiment.log` for full details.
+   * @returns The newly-created `Span`
+   */
+  startSpan(args: StartSpanArgs): Span;
```

Comment:
I think it's more consistent (and ergonomic) to make this `startSpan(name: string, args?: StartSpanArgs)`